lib/common/xpath.c

```diff
@@ -266,37 +266,65 @@ get_xpath_object(const char *xpath, xmlNode * xml_obj, int error_level)
     return result;
 }
 
-int
-pcmk__element_xpath(const xmlNode *xml, char *buffer, int offset,
-                    size_t buffer_size)
+/*!
+ * \internal
+ * \brief Get an XPath string that matches an XML element as closely as possible
+ *
+ * \param[in] xml  The XML element for which to build an XPath string
+ *
+ * \return A \p GString that matches \p xml, or \p NULL if \p xml is \p NULL.
+ *
+ * \note The caller is responsible for freeing the string using
+ *       \p g_string_free().
+ */
+GString *
+pcmk__element_xpath(const xmlNode *xml)
 {
-    const char *id = ID(xml);
+    const char *id = NULL;
+    GString *xpath = NULL;
 
-    if ((offset == 0) && (xml->parent != NULL)) {
-        offset = pcmk__element_xpath(xml->parent, buffer, offset, buffer_size);
+    // The document node has no parent and no name
+    if (xml->parent == NULL) {
+        return NULL;
     }
 
-    if(id) {
-        offset += snprintf(buffer + offset, buffer_size - offset,
-                           "/%s[@id='%s']", (const char *) xml->name, id);
-    } else if(xml->name) {
-        offset += snprintf(buffer + offset, buffer_size - offset,
-                           "/%s", (const char *) xml->name);
+    xpath = pcmk__element_xpath(xml->parent);
+    if (xpath == NULL) {
+        xpath = g_string_sized_new(256);
     }
 
-    return offset;
+    g_string_append_printf(xpath, "/%s", TYPE(xml));
+
+    id = ID(xml);
+    if (id != NULL) {
+        g_string_append_printf(xpath, "[@" XML_ATTR_ID "='%s']", id);
+    }
+
+    return xpath;
 }
 
+/*!
+ * \brief Get an XPath string that matches an XML element as closely as possible
+ *
+ * \param[in] xml  The XML element for which to build an XPath string
+ *
+ * \return A string that matches \p xml, or \p NULL if \p xml is \p NULL.
+ *
+ * \note The caller is responsible for freeing the string.
+ */
 char *
 xml_get_path(const xmlNode *xml)
 {
-    int offset = 0;
-    char buffer[PCMK__BUFFER_SIZE];
+    char *path = NULL;
+    GString *g_path = pcmk__element_xpath(xml);
 
-    if (pcmk__element_xpath(xml, buffer, offset, sizeof(buffer)) > 0) {
-        return strdup(buffer);
+    if (g_path == NULL) {
+        return NULL;
     }
-    return NULL;
+
+    path = strdup((const char *) g_path->str);
```

Comment:
Will do. FWIW I don't think `g_path->str` can be null if `g_path != NULL` (unless you manually go in and set `g_path->str` to `NULL` instead of using GLib functions).

Reply:
I mean assert the strdup() result is non-NULL. We can assume g_path->str is non-NULL

Reply:
will do